### DISCUSSION TITLE
Verify that exceptions are raised

### DISCUSSION THREAD

**Discussion by johnrichardrinehart** [theory/sqitch#699] [Q&A] [answered]
I have TRIGGERs that raise exceptions if invalid operations are attempted on UPDATE. I want my `verify` scripts to ensure that these exceptions are raised in these circumstances. How can I make sure that my `verify` scripts _pass_ when an exception is raised? Note: I also want my `verify` scripts to check that the normal, healthy flow works, too, which won't raise exceptions.

**Comment by theory** [comment, marked as answer]
Based on your previous posts, I assume you're using PostgreSQL, in which case, use a [DO](https://www.postgresql.org/docs/current/sql-do.html) statement with [error trapping](https://www.postgresql.org/docs/current/plpgsql-control-structures.html#PLPGSQL-ERROR-TRAPPING) in your verify script. Something like:

``` sql
DO $$
BEGIN
    -- SQL to trigger the exception
EXCEPTION WHEN foo_exception THEN
    -- Assuming foo_exception is the exception you expect.
    RETURN;
-- Everything else still fails
END;
$$
```

**Reply by johnrichardrinehart** [reply]
That worked perfectly! Thanks for linking the docs. I was looking for "exception handling" and "error handling". "trapping" was the word I should have used.

For future people or machines: I'm using a bit of extra sauce to ensure the code/message match:
```sql
CREATE OR REPLACE FUNCTION on_mint_update() RETURNS trigger AS '
BEGIN
RAISE EXCEPTION
	USING MESSAGE = ''updates not allowed on mint table'',
	ERRCODE = ''27000'';
END
' LANGUAGE plpgsql;
```
Hopefully that helps someone in the future.
